### PR TITLE
Show designator & inhibitor range

### PR DESCRIPTION
**Pull request by MortonPL**
* It is now possible to display range of designator and inhibitor units when in super weapon targeting mode. Each instance of player owned techno types listed in `[SuperWeapon]->SW.Designators` will display a circle with radius set in `[TechnoType]->DesignatorRange` or `Sight`.
  * In a similar manner, each instance of enemy owned techno types listed in `[SuperWeapon]->SW.Inhibitors` will display a circle with radius set in `[TechnoType]->InhibitorRange` or `Sight`.
* This feature can be disabled globally with `[AudioVisual]->ShowDesignatorRange=false` or per SuperWeaponType with `[SuperWeapon]->ShowDesignatorRange=false`.
* This feature can be toggled *by the player* (if enabled in the mod) with `ShowDesignatorRange` in `Ra2MD.ini` or with "Toggle Designator Range" hotkey in "Interface" category.

```{note}
Due to technical reasons, this feature only works for super weapons with `Range` >= 1.
```

In `rulesmd.ini`:
```ini
[AudioVisual]
ShowDesignatorRange=true    ; boolean

[SOMESW]                    ; SuperWeapon
ShowDesignatorRange=true    ; boolean
```

In `Ra2MD.ini`:
```ini
[Phobos]
ShowDesignatorRange=false             ; boolean
```

### CODE REVIEW

File: docs/User-Interface.md [hunk 1]
```diff
@@ -93,6 +93,17 @@ DigitalDisplay.Enable=false             ; boolean
 An example shape file for digits can be found on [Phobos supplementaries repo](https://github.com/Phobos-developers/PhobosSupplementaries)).
 ```
 
+### Show designator range
+
+* It is now possible to display range of designator units when in super weapon targeting mode. Each instance of player owned techno types listed in `[SuperWeapon]SW.Designators` will display a circle with radius set in `[TechnoType]DesignatorRange` or `Sight`.
```
Comment (Metadorius) [inline docs/User-Interface.md:98]:
```suggestion
* It is now possible to display range of designator units when in super weapon targeting mode. Each instance of player owned techno types listed in `[SuperWeapon]->SW.Designators` will display a circle with radius set in `[TechnoType]->DesignatorRange` or `Sight`.
```

File: docs/User-Interface.md [hunk 2]
```diff
@@ -93,6 +93,17 @@ DigitalDisplay.Enable=false             ; boolean
...
+* This feature can be enabled with `ShowDesignatorRange=true` in `Ra2MD.ini` or with "Toggle Designator Range" hotkey in "Interface" category.
```
Comment (Metadorius) [inline docs/User-Interface.md:99]: I am kinda conflicted on that one, there's no problem with that but...
1) Is that needed? I mean designator/inhibitior radius should be mandatory to show IMO, no?
2) Feels kinda weird to just have the hotkey toggle the config option?

**Comment by Starkku** [comment]
Not sure if both hotkey and user INI setting are needed for this, should probably pick one or the other. Although I guess if it is just hotkey it won't save the preference across sessions but if it is a feature where you'd want that then just having the user INI option is better anyway.

Code itself looks good, but it should really be extended to work with inhibitors too. Display the indicators in the inhibitor's owner's color. Not sure what would do when/if Suppressors/Attractors (friendly inhibitors and enemy designators) are added but that is a bridge that can be crossed then.

**Comment by Metadorius** [comment]
Considering those ranges only get displayed when you're firing a superweapon (btw does it only show relevant designators/inhibitors?) I don't think it's ever needed to be manually toggleable because it's not obstructive and it's kinda toggled automatically when firing the SW anyways.

**Comment by Starkku** [comment]
`Due to technical reasons, this feature only works for super weapons with Range >= 1.`

What technical reasons? I replaced your hook with:
```cpp
DEFINE_HOOK(0x6DBE74, Tactical_SuperLinesCircles_ShowDesignatorRange, 0x7)
{
	if (!Phobos::Config::ShowDesignatorRange || !(RulesExt::Global()->ShowDesignatorRange) || Unsorted::CurrentSWType == -1)
		return 0;

	const auto pSuperType = SuperWeaponTypeClass::Array()->GetItem(Unsorted::CurrentSWType);
	const auto pExt = SWTypeExt::ExtMap.Find(pSuperType);

	if (!pExt->ShowDesignatorRange)
		return 0;

	for (const auto pCurrentTechno : *TechnoClass::Array)
	{
		const auto pCurrentTechnoType = pCurrentTechno->GetTechnoType();
		const auto pOwner = pCurrentTechno->Owner;

		if (!pCurrentTechno->IsAlive
			|| pCurrentTechno->InLimbo
			|| !pExt->SW_Designators.Contains(pCurrentTechnoType)
			|| !((pOwner == HouseClass::CurrentPlayer)
				|| EnumFunctions::CanTargetHouse(AffectedHouse::Enemies, HouseClass::CurrentPlayer, pOwner)))
		{
			continue;
		}

		const auto pTechnoTypeExt = TechnoTypeExt::ExtMap.Find(pCurrentTechnoType);

		const float radius = pOwner == HouseClass::CurrentPlayer ?
			(float)(pTechnoTypeExt->DesignatorRange.Get(pCurrentTechnoType->Sight)) :
			(float)(pTechnoTypeExt->InhibitorRange.Get(pCurrentTechnoType->Sight));

		CoordStruct coords = pCurrentTechno->GetCenterCoords();
		coords.Z = MapClass::Instance->GetCellFloorHeight(coords);
		const auto color = pOwner->Color;
		Game::DrawRadialIndicator(false, true, coords, color, radius, false, true);
	}

	return 0;
}
```

E.g move it to the start of the function and fetch the SW type yourself. And it appears to work just fine, only issue is minor redundancy in fetching the SW type potentially twice but this comes in exchange for working independently from SW's `Range` or the target lines option. Couple of minor optimizations with the checks / their order as well. Checking extension class pointers for nullptr is generally speaking futile exercise - `ExtMap::Find` will under most circumstances only now return nullptr if you call it with nullptr, in which case you're gonna have other problems anyway (missing type) and if it is not a pointer of correct type it'll just crash or return garbage regardless. Cost of performance is security I suppose.

File: docs/User-Interface.md [hunk 3]
```diff
@@ -94,6 +94,28 @@ DigitalDisplay.Enable=false             ; boolean
 An example shape file for digits can be found on [Phobos supplementaries repo](https://github.com/Phobos-developers/PhobosSupplementaries)).
 ```
 
+### Show designator & inhibitor range
+
+* It is now possible to display range of designator and inhibitor units when in super weapon targeting mode. Each instance of player owned techno types listed in `[SuperWeapon]->SW.Designators` will display a circle with radius set in `[TechnoType]->DesignatorRange` or `Sight`.
+  * In a similar manner, each instance of enemy owned techno types listed in `[SuperWeapon]->SW.Inhibitors` will display a circle with radius set in `[TechnoType]->InhibitorRange` or `Sight`.
+* This feature can be disabled globally with `[AudioVisual]->ShowDesignatorRange=false` or per SuperWeaponType with `[SuperWeapon]->ShowDesignatorRange=false`.
+* This feature can be toggled *by the player* (if enabled in the mod) with `ShowDesignatorRange` in `Ra2MD.ini` or with "Toggle Designator Range" hotkey in "Interface" category.
```
Comment (Metadorius) [inline docs/User-Interface.md:102]:
```suggestion
- It is now possible to display range of designator and inhibitor units when in super weapon targeting mode. Each instance of player owned techno types listed in `[SuperWeapon]->SW.Designators` will display a circle with radius set in `[TechnoType]->DesignatorRange` or `Sight`.
  - In a similar manner, each instance of enemy owned techno types listed in `[SuperWeapon]->SW.Inhibitors` will display a circle with radius set in `[TechnoType]->InhibitorRange` or `Sight`.
- This feature can be disabled globally with `[AudioVisual]->ShowDesignatorRange=false` or per SuperWeaponType with `[SuperWeapon]->ShowDesignatorRange=false`.
- This feature can be toggled *by the player* (if enabled in the mod) with `ShowDesignatorRange` in `Ra2MD.ini` or with "Toggle Designator Range" hotkey in "Interface" category.
```

Comment (Metadorius) [reply docs/User-Interface.md:102]: Maybe the hotkey needs mentioning on the UI/UX page?

File: src/Ext/SWType/Hooks.cpp [hunk 4]
```diff
@@ -34,3 +34,43 @@ DEFINE_HOOK(0x6CB5EB, SuperClass_Grant_ShowTimer, 0x5)
...
+				|| EnumFunctions::CanTargetHouse(AffectedHouse::Enemies, HouseClass::CurrentPlayer, pOwner)))
+		{
+			continue;
+		}
+
+		const auto pTechnoTypeExt = TechnoTypeExt::ExtMap.Find(pCurrentTechnoType);
+
+		const float radius = pOwner == HouseClass::CurrentPlayer ?
+			(float)(pTechnoTypeExt->DesignatorRange.Get(pCurrentTechnoType->Sight)) :
+			(float)(pTechnoTypeExt->InhibitorRange.Get(pCurrentTechnoType->Sight));
```
Comment (Metadorius) [inline src/Ext/SWType/Hooks.cpp:67]:
```suggestion
		const float radius = pOwner == HouseClass::CurrentPlayer
			? (float)(pTechnoTypeExt->DesignatorRange.Get(pCurrentTechnoType->Sight))
			: (float)(pTechnoTypeExt->InhibitorRange.Get(pCurrentTechnoType->Sight));
```

Comment (Metadorius) [reply src/Ext/SWType/Hooks.cpp:67]: Operators at the beginning of the line have better readability.

**Review by Metadorius — APPROVED** [review]
LGTM after fixing

**Review by Metadorius — APPROVED** [review]

**Pull request merged**